### PR TITLE
Checks for an existing config file before overwriting.

### LINKED ISSUE #92: Bug: The latest update replaces pre-existing configuration files
It's a good thing I take backups, because the armhf .deb file appears to have completely replaced my `/etc/laminar.conf` - regardless of the fact that I've actually changed it since installation.

### PR DESCRIPTION
**Pull request by jakimfett**
Checks for the existence of an `/etc/laminar.conf` file before writing, and writes to `/etc/laminar.conf.example` should the default file already exist.

Should technically resolve #92, but might be better off standardized for all config/service/autocomplete files.

My cursory examination of the `cmake` system failed to turn up an easy-ish way to obtain user input and alter the outcome accordingly (eg, merge or copy the new/default config over top of the potentially-edited file), or even to check if the existing file is modified.

Open to suggestions on a better way to do this.  
Can't say I'm a fan of cmake so far.  

:-/

### CODE REVIEW

**Comment by ohwgiles** [comment]
Thanks for contributing! Unfortunately I have an issue with this. `make install` is more useful as a tool to collect files for package management. With this patch, if I try to prepare a laminar package *to be installed on another machine*, the package contents according to `make install` might be different depending on whether laminar is installed on the packager's machine, which is wrong. Likewise if I want to install to another path with `make DESTDIR=some/where/else install`. It must be the package manager's job to handle updated configuration files.

**Comment by jakimfett** [comment]
So, if I'm understanding correctly, the 'golden path' for installing the latest rendition of `laminar` is:
1. Obtain source
1. Build source
1. Install source to local folder
1. Use packing magic to produce a deb/rpm
1. Install the deb/rpm using system tooling

Doesn't this exclude anything that is outside the Debian and/or RedHat paradigms?

---

The bit about the files differing is noted, and kinda what I was attempting to get at with https://github.com/ohwgiles/laminar/pull/96/commits/b2e873e853a21080f7f1dfb5da0502c5f41fdddc#diff-af3b638bc2a3e6c650974192a53c7291R121, although I did a poor job of communicating that.

I was attempting to get the install step itself to check if there was a file before overwriting, but clearly this method is suboptimal, although after.

**Comment by ohwgiles** [comment]
> So, if I'm understanding correctly, the 'golden path' for installing the latest rendition of laminar is:

The best way to install software is always through your distribution's package manager. Because laminar isn't yet widely packaged, I provide quick-and-dirty `deb` and `rpm` packages for convenience, but the goal would be that someone from each distribution community takes enough interest to package it into their distribution. This is [happening now for Debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=919181).

> Doesn't this exclude anything that is outside the Debian and/or RedHat paradigms?

As above, hopefully other distros will be interested sometime. I also maintain an [AUR package](https://aur.archlinux.org/packages/laminar/)

File: CMakeLists.txt [hunk 1]
```diff
@@ -104,10 +114,21 @@ if(BUILD_TESTS)
...
+install(FILES etc/laminar.service DESTINATION ${SYSTEMD_UNITDIR})
+
+# Don't overwrite the config file if it already exists at the target location
+# @todo - migration/override script if the defaults change or the options get pruned?
+if(NOT EXISTS "${CMAKE_INSTALL_PREFIX}etc/laminar.conf")
+    install(
+        FILES etc/laminar.conf
+        DESTINATION ${CMAKE_INSTALL_PREFIX}etc
+    )
+endif(NOT EXISTS "${CMAKE_INSTALL_PREFIX}etc/laminar.conf")
```
Comment (ohwgiles) [inline CMakeLists.txt]: Again, you can't really do this. CMake runs at configuration time, and if you are creating a package to install on a different machine (which CMake cannot detect), the file should always be installed, even if the host computer has one already.

Comment (jakimfett) [reply CMakeLists.txt]: That's why I changed it to check at the `CMAKE_INSTALL_PREFIX` location. Even if the system the user is compiling on already has an `/etc/laminar.conf`, this should correctly output if the prefix tells `make` to output to somewhere other than `/etc`.

I could also be wrong, and it's _not_ possible to do both local installs and packaging cleanly, but it's late and I only hacked on this enough to un-block my progress on something else.

Barring a fire that takes my attention elsewhere, I'll poke it again tomorrow and see what I can come up with.

**Comment by jakimfett** [comment]
> The best way to install software is always through your distribution's package manager.  

Eeehhhhhhh kinda disagree on this. On several levels.  

Package manager is definitely a step up from 'binary downloaded via browser from random internet site', but for rather a lot of hosted-on-*nix application-level stuff, the package repos are out of date, sometimes dangerously so, and it's always a tossup if there's a 3rd party repo.

Having the compile process, as copypasta'd from the `README.md` instructions, overwrite destructively, _without warning_, seems odd and needing-to-be-fixed.

I'm really glad I learned about this in testing, not at 0310 some weekend when there's a security patch rolling out.

Having the compile/install process do unexpectedly destructive things while I'm trying to fix something else is not how I, as a sysadmin, want to learn about the quirks of a project's build process. 

As far as the general question of "who actually compiles instead of using packages these days", I for one compile rather a lot, it's a large part of why I'm using a continuous integration server.

My default for anything I'm actually using directly (rather than using by dependency or using obliquely in that it was part of the base system which I've yet to tinker with) is to obtain the source, and the source of dependencies, then automate the build-deploy process so that if/when a problematic thing gets patched in the source, I can single-command fix it across most/all of my infrastructure.

---
Tangentially, I've refrained from marking this PR as ready for review, since I'm very aware that I'm, to a certain extent, flailing in the dark here, especially from a perspective that doesn't include the systems I'm actually using these changes on, and without the roadmap for my next several years where I really, really want to be able to rely on a tool as useful as laminar has the potential to be.

From my perspective, if there's a way to make it work both for the local compile and for the OS specific packaging, without breaking the abstraction of simplicity for the user (whether power-user or developer or otherwise), it's always better to bake in that type of robustness.

---
tl;dr - imo parity matters, gimme a bit to see if I can come up with a workable solution.

**Comment by ohwgiles** [comment]
> for rather a lot of hosted-on-*nix application-level stuff, package repos are out of date, sometimes dangerously so

Sounds like you need different repositories. This is not so much an issue with the package manager itself. I'd really suggest spending the effort on making a package for your distro; that way you get package manager features like tracking which package owns files in your root filesystem.

> Having the compile process, as copypasta'd from the README.md instructions, overwrite destructively, without warning, seems odd and needing-to-be-fixed.

In my experience, this is standard practice. `make install` always behaves like this. Sysadmins choosing to install from source should be used to it.

> My default for anything I'm actually using directly (rather than using by dependency or using obliquely in that it was part of the base system which I've yet to tinker with) is to obtain the source, and the source of dependencies, then automate the build-deploy process

I'd suggest adding "building a distro package and installing it" to your pipeline.

--

All that said, I'm not opposed to a "safer `make install`". Maybe consider something along these lines:
```
add_custom_target(install/confirm COMMAND bash -c "t=$(mktemp -d); make DESTDIR=$t install && tar cC $t | tar xvkC /; rm -rf $t;")
```
(Untested. can't check now)

File: CMakeLists.txt (hunk 1 above)
Comment (ohwgiles) [reply CMakeLists.txt]: `CMAKE_INSTALL_PREFIX` != `DESTDIR`

`make install` installs to `$ENV{DESTDIR}/$CMAKE_INSTALL_PREFIX/$x` where `$x` is the path given to cmake's `install`

**Comment by ohwgiles** [comment]
On the balance of things, would you prefer `make install/confirm` or simply having `make install` install (and overwrite) `laminar.conf.example` instead of `laminar.conf`?

**Comment by KAction** [comment]
I'd also note that autotools, also install files unconditionally (destructively, if you want).

**Comment by jakimfett** [comment]
>  [KAction commented](https://github.com/ohwgiles/laminar/pull/96#issuecomment-616668591)
> I'd also note that autotools, also install files unconditionally (destructively, if you want).

Destructively is how I'll call it, when potential for destruction exists.

Usability flaws can persist in legacy tools for years, but that doesn't mean we shouldn't work towards a better "normal".

Just because it _can_ be dangerous, doesn't mean it _should_ be.

---
>  [ohwgiles commented](https://github.com/ohwgiles/laminar/pull/96#issuecomment-565370570)
> On the balance of things, would you prefer `make install/confirm` or simply having `make install` install (and overwrite) `laminar.conf.example` instead of `laminar.conf`?

Having install/confirm would give the user agency in what happens, but writing to `<conf>.example` by default means an extra step to keep a clean system.

It's mostly about awareness...so long as the user knows and can react to the knowledge that something (potentially) destructive is about to happen, it's enough, imo.

**Comment by ohwgiles** [comment]
> It's mostly about awareness...so long as the user knows and can react to the knowledge that something (potentially) destructive is about to happen, it's enough, imo.

Then perhaps let's just add a clear warning to the README
